### PR TITLE
Fix/tt 5563 input default

### PR DESCRIPTION
**Pull request by kofoworola**
https://tyktech.atlassian.net/browse/TT-5563

[changelog]
Add support for input field default values

### CODE REVIEW

File: pkg/ast/ast_type.go [hunk 1]
```diff
@@ -27,10 +27,24 @@ type Type struct {
 	OfType   int
 }
 
+func (d *Document) BaseTypeNameBytes(ref int) ByteSlice {
+	if kind := d.Types[ref].TypeKind; kind == TypeKindNonNull || kind == TypeKindList {
+		return d.TypeNameBytes(d.Types[ref].OfType)
```
Comment (Aenimus) [inline pkg/ast/ast_type.go:32]: So this can be infinitely nested. You know you're at the root once OfType is -1.

File: pkg/ast/ast_type.go [hunk 2]
```diff
@@ -27,10 +27,24 @@ type Type struct {
...
+	return d.TypeNameBytes(ref)
+}
+
 func (d *Document) TypeNameBytes(ref int) ByteSlice {
 	return d.Input.ByteSlice(d.Types[ref].Name)
 }
 
+func (d *Document) BaseTypeNameString(ref int) string {
+	if kind := d.Types[ref].TypeKind; kind == TypeKindNonNull || kind == TypeKindList {
+		return d.TypeNameString(d.Types[ref].OfType)
```
Comment (Aenimus) [inline pkg/ast/ast_type.go:43]: Above comment applies here too. It also means you don't need to compare the type.

File: pkg/astnormalization/inject_input_default_values.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	walker.RegisterEnterDocumentVisitor(visitor)
+	walker.RegisterVariableDefinitionVisitor(visitor)
+	return visitor
+}
+
+type inputFieldDefaultInjectionVisitor struct {
+	*astvisitor.Walker
+
+	operation  *ast.Document
+	definition *ast.Document
```
Comment (Aenimus) [inline pkg/astnormalization/inject_input_default_values.go:23]: Just making sure you need both operation definition?

Comment (kofoworola) [reply pkg/astnormalization/inject_input_default_values.go:23]: yeah need the definition to check the types

Comment (Aenimus) [reply pkg/astnormalization/inject_input_default_values.go:23]: Gotcha

File: pkg/ast/ast_type.go (hunk 1 above)
Comment (kofoworola) [reply pkg/ast/ast_type.go:32]: Oh i see, i did not know that, thanks for pointing that!

File: pkg/astnormalization/inject_input_default_values.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+			} else {
+				continue
+			}
+		}
+		if !valDef.DefaultValue.IsDefined {
+			continue
+		}
+		defVal, err := in.definition.ValueToJSON(valDef.DefaultValue.Value)
+		if err != nil {
+			in.StopWithInternalErr(err)
```
Comment (buraksezer) [inline pkg/astnormalization/inject_input_default_values.go:84]: Missing `return` or `break` statement?

File: pkg/astnormalization/inject_input_default_values.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+		}
+
+		newVariables, err := jsonparser.Set(in.operation.Input.Variables, defVal, keys...)
+		if err != nil {
+			in.StopWithInternalErr(err)
```
Comment (buraksezer) [inline pkg/astnormalization/inject_input_default_values.go:89]: Missing `return` or `break` statement?

File: pkg/astnormalization/astnormalization.go [hunk 6]
```diff
@@ -180,7 +180,12 @@ func (o *OperationNormalizer) setupOperationWalkers() {
 	if o.options.removeUnusedVariables {
 		deleteUnusedVariables(&other)
 	}
-	o.operationWalkers = append(o.operationWalkers, &fragmentInline, &extractVariablesWalker, &other)
+
+	inputDefaultInjectorWalker := astvisitor.NewWalker(48)
```
Comment (buraksezer) [inline pkg/astnormalization/astnormalization.go:184]: Why don't you just use the `other` visitor?

File: pkg/astnormalization/inject_input_default_values.go (hunk 5 above)
Comment (kofoworola) [reply pkg/astnormalization/inject_input_default_values.go:92]: Oh thanks forgot about that

File: pkg/astnormalization/astnormalization.go (hunk 6 above)
Comment (kofoworola) [reply pkg/astnormalization/astnormalization.go:184]: oh thanks for pointing that out, i'll need to write a bit more tests with other visitor edge cases to check this though

File: pkg/astnormalization/inject_input_default_values.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+		return
+	}
+	if node.Kind != ast.NodeKindInputObjectTypeDefinition {
+		return
+	}
+
+	in.recursiveInjectInputFields(node.Ref, in.variableName)
+}
+
+func (in *inputFieldDefaultInjectionVisitor) recursiveInjectInputFields(inputObjectRef int, fieldName string) {
```
Comment (spetrunin) [inline pkg/astnormalization/inject_input_default_values.go:56]: maybe call receiver `i` or `v`?

`in.Smth` - for my IMHO is a bad name for receiver cause it looks like `input` - `in`

File: pkg/astnormalization/inject_input_default_values.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+	in.variableTree = append(in.variableTree, fieldName)
```
Comment (spetrunin) [inline pkg/astnormalization/inject_input_default_values.go:57]: it is probably not `variableTree` but `jsonPath`?

File: pkg/astnormalization/inject_input_default_values.go [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+	objectDef := in.definition.InputObjectTypeDefinitions[inputObjectRef]
+	if !objectDef.HasInputFieldsDefinition {
+		return
+	}
+	for _, i := range objectDef.InputFieldsDefinition.Refs {
+		valDef := in.definition.InputValueDefinitions[i]
+		if in.definition.Types[valDef.Type].TypeKind != ast.TypeKindNonNull {
```
Comment (spetrunin) [inline pkg/astnormalization/inject_input_default_values.go:64]: not required fields could also have a default values

File: pkg/astnormalization/inject_input_default_values.go [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+			continue
+		}
+		keys := append(in.variableTree, in.definition.InputValueDefinitionNameString(i))
+		exists, err := in.variableKeyExists(keys...)
+		typeScalarEnum := in.definition.TypeIsScalar(valDef.Type, in.definition) || in.definition.TypeIsEnum(valDef.Type, in.definition)
+		if err != nil {
```
Comment (spetrunin) [inline pkg/astnormalization/inject_input_default_values.go:70]: Probably it will be better to place error-handling right after err creation

File: pkg/astnormalization/inject_input_default_values.go [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+		return
+	}
+	for _, i := range objectDef.InputFieldsDefinition.Refs {
+		valDef := in.definition.InputValueDefinitions[i]
+		if in.definition.Types[valDef.Type].TypeKind != ast.TypeKindNonNull {
+			continue
+		}
+		keys := append(in.variableTree, in.definition.InputValueDefinitionNameString(i))
+		exists, err := in.variableKeyExists(keys...)
+		typeScalarEnum := in.definition.TypeIsScalar(valDef.Type, in.definition) || in.definition.TypeIsEnum(valDef.Type, in.definition)
```
Comment (spetrunin) [inline pkg/astnormalization/inject_input_default_values.go:69]: maybe `isTypeScalarOrEnum`?

File: pkg/astnormalization/inject_input_default_values.go [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+			if node, found := in.definition.Index.FirstNodeByNameBytes(in.definition.BaseTypeNameBytes(valDef.Type)); found {
+				in.recursiveInjectInputFields(node.Ref, keys[len(keys)-1])
+			} else {
+				continue
+			}
+		}
+		if !valDef.DefaultValue.IsDefined {
+			continue
+		}
+		defVal, err := in.definition.ValueToJSON(valDef.DefaultValue.Value)
```
Comment (spetrunin) [inline pkg/astnormalization/inject_input_default_values.go:84]: even default value could contain not all fields

```
input A {
  field: b = {bOther: "some"}
}

input B {
  bWithDefault: String! = "b_default"
  bOther: String
}
```

That is why I suggested to process recursively injected json

File: pkg/astnormalization/inject_input_default_values.go [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+
+	exists, err := in.variableKeyExists(in.variableName)
+	if err != nil {
+		in.StopWithInternalErr(err)
+		return
+	}
+	if !exists {
+		return
+	}
+	typeName := in.operation.BaseTypeNameBytes(ref)
```
[the remark this answers is not in the document]
Comment (spetrunin) [reply pkg/astnormalization/inject_input_default_values.go:44]: So I believe this visitor should be somehow similar to list coercion

File: pkg/astnormalization/inject_input_default_values_test.go [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+  nested: LowerLevelInput!
+}
+
+input LowerLevelInput {
+  firstField: Int!
+  secondField: TestEnum! = ValueOne
+}
+`
+
+func TestInputDefaultValueExtraction(t *testing.T) {
```
Comment (spetrunin) [inline pkg/astnormalization/inject_input_default_values_test.go:38]: Try to add a lot more tests cases covering edge cases with partial default values

**Review by spetrunin — CHANGES_REQUESTED** [review]

File: pkg/astnormalization/inject_input_default_values.go (hunk 9 above)
Comment (kofoworola) [reply pkg/astnormalization/inject_input_default_values.go:64]: Oh just confirmed from the graphql spec that the default value should be used.

File: pkg/astnormalization/inject_input_default_values.go (hunk 10 above)
Comment (kofoworola) [reply pkg/astnormalization/inject_input_default_values.go:70]: yeah this was an oversight, thanks!

File: pkg/astnormalization/inject_input_default_values.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (kofoworola) [reply pkg/astnormalization/inject_input_default_values.go]: Yeah i was able to add support for this

File: pkg/astnormalization/inject_input_default_values.go (hunk 12 above)
Comment (kofoworola) [reply pkg/astnormalization/inject_input_default_values.go]: Added support for this

**Review by pvormste — APPROVED** [review]
